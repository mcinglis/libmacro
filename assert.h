
// Copyright 2015  Malcolm Inglis <http://minglis.id.au>
//
// This file is part of Libmacro.
//
// Libmacro is free software: you can redistribute it and/or modify it under
// the terms of the GNU Affero General Public License as published by the
// Free Software Foundation, either version 3 of the License, or (at your
// option) any later version.
//
// Libmacro is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public
// License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with Libmacro. If not, see <https://gnu.org/licenses/>.


#ifndef LIBMACRO_ASSERT_H
#define LIBMACRO_ASSERT_H


#include <assert.h>

#include <libpp/map.h>          // PP_MAP
#include <libpp/separators.h>   // PP_SEP_SEMICOLON


#ifdef NO_ASSERT
    #define ASSERT( ... )
#else

    // @public
    // Takes a variable number of scalar expressions, and `assert`s each
    // one individually. This provides better assertion error reporting than
    // techniques like `assert( e1 && e2 && ... )`.
    #define ASSERT( ... ) \
        PP_MAP( assert, PP_SEP_SEMICOLON, __VA_ARGS__ )

#endif


#endif


