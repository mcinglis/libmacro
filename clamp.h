
// Copyright 2015  Malcolm Inglis <http://minglis.id.au>
//
// This file is part of Libmacro.
//
// Libmacro is free software: you can redistribute it and/or modify it under
// the terms of the GNU Affero General Public License as published by the
// Free Software Foundation, either version 3 of the License, or (at your
// option) any later version.
//
// Libmacro is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public
// License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with Libmacro. If not, see <https://gnu.org/licenses/>.


#ifndef LIBMACRO_CLAMP_H
#define LIBMACRO_CLAMP_H


#include <limits.h>


// @public begin


#define CLAMP( X, LOWER, UPPER ) \
    ( ( ( X ) < ( LOWER ) ) ? ( LOWER ) \
    : ( ( X ) > ( UPPER ) ) ? ( UPPER ) \
                            : ( X ) )


#define CLAMP_TO_INTMAX( X ) \
    ( ( intmax_t ) CLAMP( X, ( X < 0 ) ? INTMAX_MIN : 0, INTMAX_MAX ) )


// @public end


#endif

