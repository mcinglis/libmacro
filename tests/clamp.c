
#include <stdint.h>
#include <stdio.h>

#include "../assert.h"          // ASSERT
#include "../clamp.h"           // CLAMP


int main( void )
{
    ASSERT( CLAMP( -1, 0, 10 ) == 0,
            CLAMP( 0, 0, 10 ) == 0,
            CLAMP( 1, 0, 10 ) == 1,
            CLAMP( 9, 0, 10 ) == 9,
            CLAMP( 10, 0, 10 ) == 10,
            CLAMP( 11, 0, 10 ) == 10,
            CLAMP( 5, 5, 5 ) == 5,
            CLAMP( 5, 10, 0 ) == 10 );
    printf( "`CLAMP()` tests passed!\n" );

    ASSERT( CLAMP_TO_INTMAX( 0 ) == 0,
            CLAMP_TO_INTMAX( 1 ) == 1,
            CLAMP_TO_INTMAX( -1 ) == -1,
            CLAMP_TO_INTMAX( -12398 ) == -12398,
            CLAMP_TO_INTMAX( INTMAX_MIN ) == INTMAX_MIN,
            CLAMP_TO_INTMAX( INTMAX_MAX ) == INTMAX_MAX,
            CLAMP_TO_INTMAX( UINTMAX_MAX ) == INTMAX_MAX );
    printf( "`CLAMP_TO_INTMAX()` tests passed!\n" );
}

