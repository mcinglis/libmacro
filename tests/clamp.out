`CLAMP()` tests passed!
`CLAMP_TO_INTMAX()` tests passed!
